test/dynamo/test_functions.py

```diff
@@ -938,6 +938,46 @@ def test_tensor_is_complex(x):
         else:
             return x - 1
 
+    @make_test
+    def test_tensor_is_inference(x):
+        if x.is_inference():
+            return x + 1
+        else:
+            return x - 1
+
+    def test_is_inference_recompilation(self):
+        def fn(x):
+            if x.is_inference():
+                return x + 1
+            else:
+                return x + 2
+
+        cnts = torch._dynamo.testing.CompileCounter()
+        fn = torch._dynamo.optimize(cnts, nopython=True)(fn)
+
+        x = torch.randn(2, 2)
+        dynamo_result = fn(x)
+
+        with torch.no_grad():
+            x = torch.randn(2, 2)
+            fn(x)
+
+        self.assertEqual(cnts.frame_count, 2)  # Recompile! requires_grad changed
+
+        with torch.inference_mode():
+            x = torch.randn(2, 2)
+            fn(x)
+
+        self.assertEqual(cnts.frame_count, 3)  # Recompile! inference_mode changed
```

Comment:
You're actually only testing here that the inference mode state is guarded on, not that if a tensor is an inference mode tensor or not causes a change. Move the allocation of the tensor in the mode but do the fn call outside of it to test this.

Since you are diverging the semantics of the program on the inside, you could also just check if the result equals the eager result or not.

Reply:
I think this makes sense - thanks!